cumulusci/core/source_transforms/transforms.py

```diff
@@ -308,8 +315,51 @@ def get_replace_string(self) -> str:
             )
 
 
+class FindReplaceIdAPI(str, enum.Enum):
+    REST = "rest"
+    TOOLING = "tooling"
+
+
+class FindReplaceIdSpec(FindReplaceBaseSpec):
+    replace_record_id_query: str
+    api: FindReplaceIdAPI = FindReplaceIdAPI.REST
+
+    def get_replace_string(self, context: TaskContext) -> str:
+        org = context.org_config
+
+        if self.api is FindReplaceIdAPI.REST:
+            results = org.salesforce_client.query(self.replace_record_id_query)
+        else:
+            results = org.tooling.query(self.replace_record_id_query)
+
+        if results["totalSize"] != 1:
+            raise CumulusCIException(
+                f"The find-replace query {self.replace_record_id_query} returned {results['totalSize']} results. Exactly 1 result is required"
+            )
+
+        return results["records"][0]["Id"]
+
+
+class FindReplaceCurrentUserSpec(FindReplaceBaseSpec):
+    inject_username: bool = True
+
+    def get_replace_string(self, context: TaskContext) -> str:
+        if not self.inject_username:  # pragma: no cover
+            raise CumulusCIException(
```

Comment:
💭 : I don't love that we throw an exception here when passed a validated option. Throwing a `ValidationError` isn't much better since the validation message would be less helpful than this exception message.

Replacing this it with `return context.org_config.username if self.inject_username else self.find` *could* work, but it's not clear to me whether no-op `SourceTransforms` are a thing we want.

Reply:
I ultimately don't love this either. I feel that having many variants of find-and-replace specified by `patterns` isn't very intuitive. In this case we _know_ how to retrieve the desired data to replace the `find` value with (i.e. `org_config.username`) so the user should only have to specify a `find` token. Creating new functionality as new `Transform`s that have unique `identifier` values could help this by making the `cumulusci.yml` configuration more readable.

Current:
```
deploy:
        options:
            transforms:
                - transform: find_replace
                  options:
                      patterns:
                          - find: USER_NAME
                            replace: baz
                          - find: special_string
                            replace_record_id_query: SELECT Id from User WHERE name='User User'
                            api: rest
```
Would be:
```
deploy:
        options:
            transforms:
                - transform: find_replace_current_user
                  options:
                      patterns:
                          - find: USER_NAME
                     
                - tranform: find_replace_record_id
                  options:
                      patterns:
                          - find: special_string
                            replace_record_id_query: SELECT Id from User WHERE name='User User'
```
I realize a change like this we would likely want to tackle in another story though. @davidmreed - Thoughts here?

Reply:
I see two threads here. 

First, about _this_ implementation throwing an exception when `inject_username` is `False`. Having mulled it over a bit more, I think logging a warning and returning `self.find` is probably a better UX. For example, a user that has a username that is valid for a deployment to a packaging org who wants to replace it when deploying to a scratch org could have `inject_username: True` on some flows but not others.

---

Second, about the intuitiveness of `patterns`:

>  In this case we know how to retrieve the desired data to replace the find value with (i.e. org_config.username) so the user should only have to specify a find token.

That's a great point, Brandon. Ultimately there are two classes of substitutions:

1. Arbitrary user-defined strings (`'foo' -> 'bar'`)
2. Known metadata with a token we define (`%%%NAMESPACE%%%`)

Username and ID fall into the second category, IMO. In other words, we'd define a token and replace that (`$USERNAME$ -> cae8ou@example.com`). Being able to use the token in arbitrary substitutions would be a nice bonus:

```yaml
patterns:
  - find: username@production.org
    replace: $USERNAME$
```

We could define tokens for other OrgConfig values (org_id, etc) too.

Agreed that any change would be in another story.

Reply:
Noting this for future reference:

We also discussed several options on the `deploy` task that seem like they would be better suited to be configured under the `transforms` heading. These include things options like:
- `--unmanaged`
- `--namepace_inject`
- `--namespace_strip`
- `--namespaced_org`
- `--clean_meta_xml`

Reply:
I am fine with either the use of a different find-replace options model, or a completely separate identifier.

We have a planned story to add a `find` variant where the find-element is an XPath locator instead of an arbitrary token, which should allow all of these transforms to be Metecho-compatible.

The built-in transformations that `deploy` provides have already been implemented under the hood as transform classes. Refactoring how we parse the options to use transforms instead is a pretty major breaking change, so I haven't pushed for it yet. We could introduce that as an optional next-generation config.